### PR TITLE
doc: add documents for runtime.

### PR DESCRIPTION
**Pull request by FeelyChau**

### CODE REVIEW

File: packages/core/src/runtime.ts [hunk 1]
```diff
@@ -1,82 +1,42 @@
...
-/**
- * progress infomation
+ * The model script can emit the training progress through the API `Runtime.notifyProgress`.
  */
 export interface ProgressInfo {
-  // progress percentage, 0 - 100
+  /**
+   * The training progress percentage, it should be [0, 100].
+   */
   progressValue: number;
```
Comment (yorkie) [inline packages/core/src/runtime.ts:10]:
```suggestion
  value: number;
```

File: packages/costa/src/index.ts [hunk 2]
```diff
@@ -18,19 +15,20 @@ import * as dataCook from '@pipcook/datacook';
 import * as path from 'path';
 import Debug from 'debug';
 import { importFrom } from './utils';
+
 const debug = Debug('costa.runnable');
 
 export * from './types';
 
-export type DefaultRuntime = Runtime<Sample<any>, DatasetMeta>;
+export type DefaultRuntime = Runtime<dataCook.Dataset.Types.Sample<any>, dataCook.Dataset.Types.DatasetMeta>;
```
Comment (yorkie) [inline packages/costa/src/index.ts:23]: dataCook or Datacook?

File: packages/costa/src/index.ts [hunk 3]
```diff
@@ -46,28 +44,39 @@ export interface PipelineWorkSpace {
...
-  // framework directory
+  /**
+   * framework directory
+   */
   frameworkDir: string;
 }
 
-export interface PipelineRunnerOption {
+/**
+ * `Costa` constructor options. For `Costa` creatation,
```
Comment (yorkie) [inline packages/costa/src/index.ts:54]:
```suggestion
 * `Costa` constructor options. For `Costa` creation,
```

File: packages/costa/src/index.ts [hunk 4]
```diff
@@ -46,28 +44,39 @@ export interface PipelineWorkSpace {
...
+ * the runtime should pass the framework information and specific
+ * the workspace directory paths.
```
Comment (yorkie) [inline packages/costa/src/index.ts:56]:
```suggestion
 * the specific workspace directory paths.
```

File: packages/costa/src/index.ts [hunk 5]
```diff
@@ -46,28 +44,39 @@ export interface PipelineWorkSpace {
...
 }
 
 /**
- * The pipeline runner executes the scripts in pipeline
+ * The pipeline runner who executes the scripts in the pipeline.
  */
 export class Costa {
   /**
-   * the framework directroy
+   * The context of the pipeline instance.
```
Comment (yorkie) [inline packages/costa/src/index.ts:68]:
```suggestion
   * The context of the pipeline script.
```

File: packages/costa/src/index.ts [hunk 6]
```diff
@@ -46,28 +44,39 @@ export interface PipelineWorkSpace {
...
    */
   public context: ScriptContext;
+
   /**
-   * Create a runnable by the given runtime.
+   * The consturctor of Costa.
```
Comment (yorkie) [inline packages/costa/src/index.ts:73]:
```suggestion
   * The consturctor of `Costa`.
```

File: packages/costa/src/index.ts [hunk 7]
```diff
@@ -46,28 +44,39 @@ export interface PipelineWorkSpace {
...
    */
   constructor(
-    public options: PipelineRunnerOption
+    /**
+     * The options for the costa, includes the framework info and the workspace directory paths.
```
Comment (yorkie) [inline packages/costa/src/index.ts:77]:
```suggestion
     * The options for the `Costa`, includes the framework info and the workspace directory paths.
```

File: packages/costa/src/index.ts [hunk 8]
```diff
@@ -92,24 +101,26 @@ export class Costa {
...
+      throw new TypeError(`no entry found in ${script.name}(${script.path})`);
     }
     return fn;
   }
 
   /**
-   * start datasource script.
-   * @param script the metadata of script
-   * @param options options of the pipeline
+   * Start datasource script.
```
Comment (yorkie) [inline packages/costa/src/index.ts:120]:
```suggestion
   * Run a data source script.
```

File: packages/costa/src/types.ts [hunk 9]
```diff
@@ -1,30 +1,60 @@
 import { ParsedUrlQuery } from 'querystring';
 
 /**
- * enum of script type
+ * There are three types of Pipcook script: `DataSource`, `Dataflow`, `Model`.
  */
-export enum ScriptType { DataSource, Dataflow, Model }
+export enum ScriptType {
+  /**
+   * The `DataSource` script needs to collect all the origin data,
```
Comment (yorkie) [inline packages/costa/src/types.ts:8]:
```suggestion
   * The `DataSource` script is to collect data,
```

File: packages/costa/src/types.ts [hunk 10]
```diff
@@ -1,30 +1,60 @@
...
+   * and offers an API object for data access.
```
Comment (yorkie) [inline packages/costa/src/types.ts:9]:
```suggestion
   * and offers an API object for the following flow.
```

File: packages/costa/src/types.ts [hunk 11]
```diff
@@ -1,30 +1,60 @@
...
+   */
+  DataSource,
+  /**
+   * The `Dataflow` script processes the origin data form `DataSource` script.
```
Comment (yorkie) [inline packages/costa/src/types.ts:13]:
```suggestion
   * The `Dataflow` script processes the original data from `DataSource` script.
```

File: packages/costa/src/types.ts [hunk 12]
```diff
@@ -1,30 +1,60 @@
...
- * enum of script type
+ * There are three types of Pipcook script: `DataSource`, `Dataflow`, `Model`.
  */
-export enum ScriptType { DataSource, Dataflow, Model }
+export enum ScriptType {
+  /**
+   * The `DataSource` script needs to collect all the origin data,
+   * and offers an API object for data access.
+   */
+  DataSource,
```
Comment (yorkie) [inline packages/costa/src/types.ts:11]: DataSource/DataFlow vs. Datasource/Dataflow?

File: packages/costa/src/types.ts [hunk 13]
```diff
@@ -1,30 +1,60 @@
...
+  /**
+   * The `Dataflow` script processes the origin data form `DataSource` script.
+   */
+  Dataflow,
+  /**
+   * The `Model` script accesses the data from `DataSource` and `Dataflow` sciprts,
```
Comment (yorkie) [inline packages/costa/src/types.ts:17]:
```suggestion
   * The `Model` script is to access the data from `DataSource` and `Dataflow` scripts,
```

File: packages/costa/src/types.ts [hunk 14]
```diff
@@ -1,30 +1,60 @@
...
+   * and trains the model.
```
Comment (yorkie) [inline packages/costa/src/types.ts:18]:
```suggestion
   * and to train the model optionally.
```

File: packages/costa/src/types.ts [hunk 15]
```diff
@@ -1,30 +1,60 @@
...
+   */
+  Model
+}
 
 /**
- * pipcook script stucture
+ * The pipcook script stucture. It describes a scrpt and tell `Costa` how to run this script.
```
Comment (yorkie) [inline packages/costa/src/types.ts:24]:
```suggestion
 * The Pipcook script structure. It describes a script and tells `Costa` how to run.
```

File: packages/costa/src/types.ts [hunk 16]
```diff
@@ -1,30 +1,60 @@
...
-export enum PackageType { Python = 'python', JS = 'js' }
+export enum PackageType {
+  /**
+   * Python package which can be used through `boa.import`.
+   */
+  Python = 'python',
+  /**
+   * JS module which can be used through `import`.
+   */
+  JS = 'js'
```
Comment (yorkie) [inline packages/costa/src/types.ts:56]: use mjs?

File: packages/costa/src/types.ts [hunk 17]
```diff
@@ -33,27 +63,49 @@ export enum PackageType { Python = 'python', JS = 'js' }
...
  */
 export interface PipcookFramework {
-  // the location of the framework
+  /**
+   * The location of the framework.
+   */
   path: string;
-  // framework name
+  /**
+   * Framework name.
```
Comment (yorkie) [inline packages/costa/src/types.ts:71]:
```suggestion
   * The framework name.
```

File: packages/costa/src/types.ts [hunk 18]
```diff
@@ -33,27 +63,49 @@ export enum PackageType { Python = 'python', JS = 'js' }
...
+   */
   pythonVersion: string | null;
-  // python site-packages path in the directory, 'site-packages' by default.
+  /**
+   * Python site-packages relative path in the directory, 'site-packages' by default.
+   */
   pythonPackagePath: string | null;
-  // js node modules path in the directory, 'node_modules' by default.
+  /**
+   * Node modules relative path in the directory, 'node_modules' by default.
```
Comment (yorkie) [inline packages/costa/src/types.ts:107]: Ditto.

File: tools/mkdoc.sh [hunk 19]
```diff
@@ -10,4 +10,5 @@ function mkdoc() {
     --mode file
 }
 
-mkdoc docs/typedoc core "Pipcook Interfaces"
+mkdoc docs/typedoc/script core "Pipcook Interfaces"
+mkdoc docs/typedoc/runtime costa "Pipcook Runtime Interfaces"
```
Comment (yorkie) [inline tools/mkdoc.sh:14]:
```suggestion
mkdoc docs/typedoc/runtime costa "Costa Interfaces"
```

Comment (yorkie) [reply tools/mkdoc.sh:14]: Or use the package name?

**Review by yorkie — CHANGES_REQUESTED** [review]

File: packages/costa/src/index.ts (hunk 2 above)
Comment (FeelyChau) [reply packages/costa/src/index.ts:23]: @WenheLI What's your opinion？

Comment (yorkie) [reply packages/costa/src/index.ts:23]: See https://github.com/alibaba/pipcook/pull/739/files#diff-731eeb09ecd79fd1c3157ce640aa046c8c798d2fdf133837b696e4e1bfa8c0c4R1

@FeelyChau we need a consistent name for the same package.

File: tools/mkdoc.sh (hunk 19 above)
Comment (FeelyChau) [reply tools/mkdoc.sh:14]: `Core` and `Costa`? It looks more simpler.

**Review by yorkie — APPROVED** [review]

**Review by yorkie — APPROVED** [review]

File: packages/costa/src/types.ts (hunk 12 above)
Comment (WenheLI) [reply packages/costa/src/types.ts:11]: Dataflow could be one word; while `DataSource` usually refers to a combination of two words.

**Review by WenheLI — APPROVED** [review]

File: packages/costa/src/index.ts (hunk 2 above)
Comment (FeelyChau) [reply packages/costa/src/index.ts]: Should be `Datacook`.

**Review by rickycao-qy — APPROVED** [review]
LGTM

**Pull request merged**